packages/http-client-java/generator/http-client-generator-core/src/main/java/com/microsoft/typespec/http/client/generator/core/template/ClientMethodTemplate.java

```diff
@@ -65,54 +66,45 @@ public static ClientMethodTemplate getInstance() {
-            // 1. Assumes that the expression is nullable.
-            // 2. Assumes that the client method returns a reactive response.
-            // 3. Assumes that the reactive response is a Mono.
-            JavaIfBlock nullCheck = function.ifBlock(expressionToCheck + " == null", ifBlock -> {
-                if (JavaSettings.getInstance().isSyncStackEnabled() && clientMethodType.isSync()) {
+        final boolean isSync = JavaSettings.getInstance().isSyncStackEnabled() && clientMethod.getType().isSync();
+        // Parameter expressions to validate for non-null value.
+        final List<String> paramReferenceExpressions = clientMethod.getRequiredNullableParameterExpressions();
+        // Parameter expressions for custom validation (key is the expression, value is the validation).
+        final Map<String, String> validateParamExpressions = new HashMap<>(clientMethod.getValidateExpressions());
+
+        for (String paramReferenceExpression : paramReferenceExpressions) {
+            final JavaIfBlock nullCheck = function.ifBlock(paramReferenceExpression + " == null", ifBlock -> {
+                final String paramRequiredException = "new IllegalArgumentException(\"Parameter "
+                    + paramReferenceExpression + " is required and cannot be null.\")";
+                if (isSync) {
                     if (settings.isUseClientLogger()) {
-                        ifBlock.line("throw LOGGER.atError().log(" + exceptionExpression + ");");
+                        ifBlock.line("throw LOGGER.atError().log(" + paramRequiredException + ");");
                     } else {
-                        ifBlock.line("throw " + exceptionExpression + ";");
+                        ifBlock.line("throw " + paramRequiredException + ";");
                     }
                 } else {
-                    ifBlock.methodReturn("Mono.error(" + exceptionExpression + ")");
+                    ifBlock.methodReturn("Mono.error(" + paramRequiredException + ")");
                 }
             });
 
-            String potentialValidateExpression = validateExpressions.get(expressionToCheck);
-            if (potentialValidateExpression != null) {
-                nullCheck.elseBlock(elseBlock -> elseBlock.line(potentialValidateExpression + ";"));
+            final String validateParamExpression = validateParamExpressions.remove(paramReferenceExpression);
+            if (validateParamExpression != null) {
+                nullCheck.elseBlock(elseBlock -> elseBlock.line(validateParamExpression + ";"));
             }
         }
 
-        for (Map.Entry<String, String> validateExpression : validateExpressions.entrySet()) {
-            if (!expressionsToCheckSet.contains(validateExpression.getKey())) {
-                function.ifBlock(validateExpression.getKey() + " != null",
-                    ifBlock -> ifBlock.line(validateExpression.getValue() + ";"));
-            }
```

Comment:
not sure whether there could be impact from this condition.

Reply:
We now use a `HashMap` (validateParamExpressions) from which already validated expressions are removed during the first iteration. So, the remaining expressions those have not yet been validated are handled here. Hence there will be no change in how the generated code looks like.